include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+ *         assuming that the layout describes an image
+ */
+struct ImageLayoutInfo : LayoutInfo {
+  DALI_HOST_DEV
+  static int NumSpatialDims(const TensorLayout &tl) {
+    int s = 0;
+    for (int i = 0; i < tl.ndim(); i++) {
+      switch (tl[i]) {
+      case 'D':
+      case 'H':
+      case 'W':
+        s++;
+        break;
+      default:
+        break;
+      }
+    }
+    return s;
+  }
+  DALI_HOST_DEV
+  static int ChannelDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'C');
+  }
+  DALI_HOST_DEV
+  static bool HasChannel(const TensorLayout &tl) {
+    return ChannelDimIndex(tl) >= 0;
+  }
+  DALI_HOST_DEV
+  static bool Is2D(const TensorLayout &tl) {
+    return NumSpatialDims(tl) == 2;
+  }
+  DALI_HOST_DEV
+  static bool Is3D(const TensorLayout &tl) {
+    return NumSpatialDims(tl) == 3;
+  }
+  DALI_HOST_DEV
+  static bool IsChannelFirst(const TensorLayout &tl) {
+    return tl[0] == 'C' || (tl[0] == 'N' && tl[1] == 'C');
+  }
+  DALI_HOST_DEV
+  static bool IsChannelLast(const TensorLayout &tl) {
+    return !tl.empty() && tl[tl.ndim()-1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsImage(const TensorLayout &tl) {
+    return NumSpatialDims(tl) >= 2;
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes a video
+ */
+struct VideoLayoutInfo : ImageLayoutInfo {
+  /** @brief Returns the index of the dimension referring to frames */
+  DALI_HOST_DEV
+  static int FrameDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'F');
+  }
+  DALI_HOST_DEV
+  static bool IsChannelFirst(const TensorLayout &tl) {
+    return tl[FrameDimIndex(tl)+1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsSequence(const TensorLayout &tl) {
+    return tl.contains('F');
+  }
+  DALI_HOST_DEV
+  static bool IsVideo(const TensorLayout &tl) {
+    return IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static bool IsStillImage(const TensorLayout &tl) {
+    return !IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetFrameLayout(const TensorLayout &tl) {
+    return tl.skip('F');
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetSequenceLayout(const TensorLayout &tl) {
+    if (tl.contains('F'))
+      return tl;
+    if (tl[0] == 'N')
+      return "NF" + tl.sub(1);
+    else
+      return "F" + tl;
+  }
+};
+
+/** @brief Calculates mapping of dimensions from given output to input.
+ *
+ * Array element corresponding to given output dimensions is the index of that
+ * dimension in the input layout.
+ *
+ * Example:
+ * ```
+ * input_layout = "NHWC"
+ * output_layout = "NCHW"
+ * result = { 0, 3, 1, 2 }
+ * ```
+ * Explanation:
+ *   - 'N' appears at index 0 in both input_layout and output_layout.
+ *   - 'C' appears at index 3 in input
+ *   - 'H' appears at index 1 in input
+ *   - 'W' appears at index 2 in input
+ *
+ * If there are repetitions, the repeated names are listed in increasing order, e.g.
```

Comment:
What is an example of well formed layout with repetitions?

Reply:
I've no idea - but I think that if we outright forbid it, use case will appear next week ;)